File: C++/reverse-vowels-of-a-string.cpp
class Solution {
public:
    string reverseVowels(string s) {
        vector<char> vowels;
        
        for (char c: s) {
            switch(c) {
                case 'a':
                case 'e':
                case 'i':
                case 'o':
                case 'u':
                case 'A':
                case 'E':
                case 'I':
                case 'O':
                case 'U':                    
                    vowels.push_back(c);
                    break;
            }
        }
        
        string t;
        for (char c: s) {
            switch(c) {
                case 'a':
                case 'e':
                case 'i':
                case 'o':
                case 'u':
                case 'A':
                case 'E':
                case 'I':
                case 'O':
                case 'U':                    
                    t += vowels.back();
                    vowels.pop_back();
                    break;
                default:
                    t += c;
            }            
        }
        return t;
    }
};

File: C#/reverse-vowels-of-a-string.cs
public class Solution {
    public string ReverseVowels(string s) {
        var vowels = new Stack<char>();
        foreach (char c in s)
        {
            if (IsVowel(c))
            {
                vowels.Push(c);
            }
        }
        var sb = new StringBuilder();
        foreach (char c in s)
        {
            if (IsVowel(c))
            {
                sb.Append(vowels.Pop());
            }
            else
            {
                sb.Append(c);
            }
        }
        return sb.ToString();
    }

    private static bool IsVowel(char c)
    {
        return c == 'a' || c == 'A'
            || c == 'e' || c == 'E'
            || c == 'i' || c == 'I'
            || c == 'o' || c == 'O'
            || c == 'u' || c == 'U';
    }
}